android/src/com/mapswithme/maps/location/SocketWrapper.java

```diff
@@ -0,0 +1,211 @@
+ * sending/receiving the raw data trough platform socket interface.
+ * <p>
+ * The instance of this class is supposed to be created in JNI layer
+ * and supposed to be used in the thread safe environment, i.e. thread safety
+ * should be provided externally (by the client of this class).
+ * <p>
+ * <b>All public methods are blocking and shouldn't be called from the main thread.</b>
+ */
+class SocketWrapper implements PlatformSocket
+{
+  private final static Logger sLogger = new DebugLogger(SocketWrapper.class.getSimpleName());
+  private final static int SYSTEM_SOCKET_TIMEOUT = 30 * 1000;
+  @Nullable
+  private Socket mSocket;
+  @Nullable
+  private String mHost;
+  private int mPort;
+  private int mTimeout;
+
+  @Override
+  public boolean open(@NonNull String host, int port)
+  {
+    if (mSocket != null)
+    {
+      sLogger.e("Socket is already opened. Seems that it wasn't closed.");
+      return false;
+    }
+
+    mHost = host;
+    mPort = port;
+
+    Socket socket = createSocket(host, port, true);
+    if (socket != null && socket.isConnected())
+    {
+      mSocket = socket;
+    }
+
+    return mSocket != null;
+  }
+
+  @Nullable
+  private static Socket createSocket(@NonNull String host, int port, boolean ssl)
+  {
+    if (ssl)
+    {
+      SocketFactory sf = SSLSocketFactory.getDefault();
+      try
+      {
+        return sf.createSocket(host, port);
+      } catch (IOException e)
+      {
+        sLogger.e("Failed to create the ssl socket, mHost", host, " mPort = ", port, e);
+      }
+    } else
+    {
+      try
+      {
+        return new Socket(host, port);
+      } catch (IOException e)
+      {
+        sLogger.e("Failed to create the socket, mHost = ", host, " mPort = ", port, e);
+      }
+    }
+
+    return null;
+  }
+
+  @Override
+  public void close()
+  {
+    if (mSocket == null)
+    {
+      sLogger.e("Socket is already closed or it wasn't opened before");
+      return;
+    }
+
+    try
+    {
+      mSocket.close();
+    } catch (IOException e)
+    {
+      sLogger.e("Failed to close socket: ", this, e);
+    } finally
+    {
+      mSocket = null;
+    }
+  }
+
+  @Override
+  public boolean read(@NonNull byte[] data, int count)
+  {
+    if (mSocket == null)
+    {
+      sLogger.e("Socket must be opened before reading");
+      return false;
+    }
+
+    sLogger.d("Read method has started, data.length = " + data.length, ", count = " + count);
+    long startTime = System.currentTimeMillis();
+    int readBytes = 0;
+    try
+    {
+      InputStream in = mSocket.getInputStream();
+      while (count != 0 && (System.currentTimeMillis() - startTime) < mTimeout)
+      {
+        try
+        {
+          sLogger.d("Attempting to read ", count, " bytes from offset = ", readBytes);
+          int read = in.read(data, readBytes, count);
+
+          if (read == -1)
+          {
+            //TODO: end of socket stream!? This moment needs to be investigated more
```

Comment:
Discussed. I will break loop in this case and the `false` value will be returned from the method because `readBytes == count` will be `false`